pinot-controller/src/main/java/com/linkedin/pinot/controller/helix/core/realtime/PinotLLCRealtimeSegmentManager.java

```diff
@@ -1199,20 +1232,31 @@ public void updateKafkaPartitionsIfNecessary(TableConfig tableConfig) {
 
     // Generate new kafka partition assignment and update the znode
     if (currentInstances.size() < currentReplicaCount) {
-      LOGGER.error("Cannot have {} replicas in {} instances for {}.Not updating partition assignment", currentReplicaCount, currentInstances.size(), realtimeTableName);
+      LOGGER.error("Cannot have {} replicas in {} instances for {}.Not updating partition assignment",
+          currentReplicaCount, currentInstances.size(), realtimeTableName);
       long numOfInstancesNeeded = currentReplicaCount - currentInstances.size();
-      _controllerMetrics.setValueOfTableGauge(realtimeTableName, ControllerGauge.SHORT_OF_LIVE_INSTANCES, numOfInstancesNeeded);
+      _controllerMetrics.setValueOfTableGauge(realtimeTableName, ControllerGauge.SHORT_OF_LIVE_INSTANCES,
+          numOfInstancesNeeded);
       return;
     } else {
       _controllerMetrics.setValueOfTableGauge(realtimeTableName, ControllerGauge.SHORT_OF_LIVE_INSTANCES, 0);
     }
-    ZNRecord newPartitionAssignment = generatePartitionAssignment(tableConfig, kafkaStreamMetadata.getKafkaTopicName(), currentPartitionCount, currentInstances);
-    writeKafkaPartitionAssignment(realtimeTableName, newPartitionAssignment);
+
+    Map<String, ZNRecord> newPartitionAssignment = generatePartitionAssignment(tableConfig, currentInstances);
+    writeKafkaPartitionAssignment(newPartitionAssignment);
```

Comment:
What happens when one kafka partition change is detected in the master controller, and vlauidation manager is updating a bunch of znodes, and a table gets added in another controller, and it tries to update the same set of znodes? What happens if a controller fails after updating some znodes and not others?

Reply:
Added a comment to fix this.